server/map_tree_updater.go

```diff
@@ -38,114 +38,135 @@ type mapTreeUpdater struct {
 
 // update updates the sparse Merkle tree at the passed-in revision with the
 // given leaf updates, and writes it to the storage. Returns the new root hash.
-func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, hkv []merkle.HashKeyValue, rev int64) ([]byte, error) {
-	// Work around a performance issue when using the map in
-	// single-transaction mode by preloading all the nodes we know the
-	// sparse Merkle writer is going to need.
+// Requires updates to be non-empty.
+func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, upd []smt.NodeUpdate, writeRev int64) ([]byte, error) {
+	// Work around a performance issue when using the map in single-transaction
+	// mode by preloading all the nodes we know the Writers are going to need.
 	if t.singleTX && t.preload {
-		if err := doPreload(ctx, tx, t.hasher.BitLen(), hkv, rev-1); err != nil {
+		if err := doPreload(ctx, tx, uint(t.hasher.BitLen()), upd, writeRev-1); err != nil {
 			return nil, err
 		}
 	}
 
-	smtWriter, err := merkle.NewSparseMerkleTreeWriter(ctx, t.tree.TreeId, rev, t.hasher, t.newTXRunner(t.tree, tx))
+	// TODO(pavelkalinnikov): Make the layout configurable.
+	w := smt.NewWriter(t.tree.TreeId, t.hasher, uint(t.hasher.BitLen()), 8)
+	shards, err := w.Split(upd)
 	if err != nil {
 		return nil, err
 	}
 
-	if err = smtWriter.SetLeaves(ctx, hkv); err != nil {
+	runTX := t.newTXFunc(tx)
+	update := func(ctx context.Context, upd []smt.NodeUpdate) (smt.NodeUpdate, error) {
```

Comment:
This deserves a comment about writing one shard
Return vars could also be named, this would save a few var declarations